### PR TITLE
python/cast_mode: Fix integer overflow of LONG_MAX+1.

### PR DESCRIPTION
**Pull request by degasus**
LONG_MAX cannot be represented as double if sizeof(long)=8, so it is usually rounded to LONG_MAX+1. So SWIG_CanCastAsInteger returned true for LONG_MAX+1 and the next cast ended up in an integer overflow.

Worse for unsigned long: It is called by smaller unsigned integers with range checks after the convertion to unsigned long. As the integer overflow yields a 0x0, this is valid for all small unsigned integers. Smaller signed integers will always fail this range check after the integer overflow.

This fixes the tests on my machine:
```
  as_l(lmaxd + LSB)
  as_ll(llmaxd + LSB)
  as_ul(ulmax + LSB)
  as_ull(ullmax + LSB)
```

Wrong old results before this commit:
```
  as_l(float(2**63)) := -0x8000000000000000
  as_ll(float(2**63)) := -0x8000000000000000
  as_uc(float(2**64)) := 0x0
  as_us(float(2**64)) := 0x0
  as_ui(float(2**64)) := 0x0
  as_ul(float(2**64)) := 0x0
  as_ull(float(2**64)) := 0x0
```

Now all of them throw an exception.

Also add tests for the min/max values of this convertions including negative tests for min-1/max+1.

### CODE REVIEW

File: Examples/test-suite/python/python_overload_simple_cast_runme.py [hunk 1]
```diff
@@ -164,34 +164,120 @@ def __float__(self):
...
+if as_l(lmaxd) != lmaxd:
+    raise RuntimeError("as_l(lmaxd)")
+
+exceptTypeError(as_l, lmin - 1, "as_l(lmin - 1)")
+exceptTypeError(as_l, lmax + 1, "as_l(lmax + 1)")
+exceptTypeError(as_l, nextafter(lmind, -inf), "as_l(lmind - LSB)")
+exceptTypeError(as_l, nextafter(lmaxd, inf), "as_l(lmaxd + LSB)")
+
 # unsigned long long
-ullmax = 9223372036854775807  # 0xffffffffffffffff
```
Comment (degasus) [inline Examples/test-suite/python/python_overload_simple_cast_runme.py:168]: Note: This does NOT match. The old value `9223372036854775807 == 2**63-1` conflicts with the comment.

File: Lib/python/pyprimtypes.swg [hunk 2]
```diff
@@ -104,7 +104,9 @@ SWIG_AsVal_dec(long)(PyObject *obj, long* val)
     if (!dispatch) {
       double d;
       int res = SWIG_AddCast(SWIG_AsVal(double)(obj,&d));
-      if (SWIG_IsOK(res) && SWIG_CanCastAsInteger(&d, LONG_MIN, LONG_MAX)) {
+      // largest double not larger than LONG_MAX
+      const double long_max = sizeof(long) == 8 ? 0x7ffffffffffffc00LL : LONG_MAX;
```
Comment (degasus) [inline Lib/python/pyprimtypes.swg:108]: I really hate hardcoding this two constants. Howver I was not able to generate this constant in C with less than 10 lines of code without any a-prioi-knowledge. I anyone is aware of a method to convert a 64 bit integer > 2**53 to double with a given rounding behavior, I will happily update this patch.

How to generate:
  C: nextafter(uint128_t(LONG_MAX)+1, 0)
  Python: hex(int(nextafter(2**63, 0)))

File: Examples/test-suite/overload_simple.i [hunk 3]
```diff
@@ -166,10 +166,17 @@ char *fid(double, double) {
...
-unsigned long long ull() { return 0ULL; }
-unsigned long long ull(unsigned long long ull) { return ull; }
-long long ll() { return 0LL; }
-long long ll(long long ull) { return ull; }
+int sizeof_long() { return sizeof(long); }
+
+unsigned long as_ul() { return 0UL; }
+unsigned long as_ul(unsigned long ul) { return ul; }
+long as_l() { return 0L; }
+long as_l(long l) { return l; }
```
Comment (degasus) [inline Examples/test-suite/overload_simple.i:174]: Just `l` was not allowed because of https://www.flake8rules.com/rules/E743.html , so I've renamed it to as_l.

File: Examples/test-suite/python/python_overload_simple_cast_runme.py [hunk 4]
```diff
@@ -164,34 +164,120 @@ def __float__(self):
...
+    raise RuntimeError("as_ul(ulmax)")
+if as_ul(ulmind) != ulmind:
+    raise RuntimeError("as_ul(ulmind)")
+if as_ul(ulmaxd) != ulmaxd:
+    raise RuntimeError("as_ul(ulmaxd)")
+
+exceptTypeError(as_ul, ulmin - 1, "as_ul(ulmin - 1)")
+exceptTypeError(as_ul, ulmax + 1, "as_ul(ulmax + 1)")
+exceptTypeError(as_ul, nextafter(ulmind, -inf), "as_ul(ulmind - LSB)")
+exceptTypeError(as_ul, nextafter(ulmind, inf), "as_ul(ulmaxd + LSB)")
```
Comment (degasus) [inline Examples/test-suite/python/python_overload_simple_cast_runme.py:210]: ulmaxd instead of ulmind, also with as_ull

**Comment by degasus** [comment]
@wsfulton This patch passed the CI, after rebase on top of your commits, it now fails. As master has failed the CI as well, I doubt there is a need for me to invastigate further.

By the way, is there something you want me to do here? It fixes an obvious bug without touching the behavior elsewhere. And it increases the test coverage.

File: Lib/python/pyprimtypes.swg (hunk 2 above)
Comment (ojwb) [reply Lib/python/pyprimtypes.swg:108]: This also seems to assume IEEE double, which is pretty much ubiquitous in current platforms but not actually required by the language standards.

It does seem fiddly to calculate though.

File: CHANGES.current [hunk 5]
```diff
@@ -7,6 +7,17 @@ the issue number to the end of the URL: https://github.com/swig/swig/issues/
 Version 4.2.0 (in progress)
 ===========================
 
+2023-XX-XX: degasus
+            Fix integer overflow / undefined behavior in the python cast_mode on sizeof(long)==8
+            plattforms for the implicit convertion of the edge cases:
```
Comment (ojwb) [inline CHANGES.current:12]: Typo: "plattforms"

**Comment by ojwb** [comment]
BTW, please don't rebase and push your PRs so much - every time you do it kicks off 100 or so builds, which is a lot of CI time to be burning when a patch is really just waiting for someone to get to it.  It can also delay CI runs for things people are actively working on since GHA limits us to fewer concurrent jobs than we kick off per push.

File: Lib/python/pyprimtypes.swg (hunk 2 above)
Comment (ojwb) [reply Lib/python/pyprimtypes.swg:108]: If we have `uint128_t` (or if `sizeof(long) == 4` and we have `uint64_t`) and we're building as C99 or have POSIX functions then we could just use `nextafter((uint128_t)LONG_MAX+1, 0)` (or `nextafter((uint64_t)LONG_MAX+1, 0)`) - checking on godbolt it seems GCC and MSVC both evaluate this at compile time (though clang doesn't seem to currently).  We could also check `FLT_RADIX` and `DBL_MANT_DIG` match IEEE and use a precomputed value for common platforms too.

Many platforms lack `uint128_t` though, and I'd guess a platform not using IEEE floating point is probably especially unlikely to.

Some other assumptions would be easy to fix though, e.g.:

```
llmaxd = float(2**53) # 64 bit double significand
```

We could add C helpers to return `FLT_RADIX` and `DBL_MANT_DIG` (from `<float.h>`) and replace this with something like:

```
llmaxd = float(flt_radix()**dbl_mant_dig()) # 64 bit double significand
```

I'm not sure if people would actually be trying to use SWIG on a platform with non-IEEE floating point though.

@wsfulton Thoughts?

**Comment by degasus** [comment]
As this is a bug fix, what do you think about adding a milestone to it?

If there is anything I can do to help on the decision about hard coding the IEEE float behavior, please let me know.

File: CHANGES.current (hunk 5 above)
Comment (ojwb) [reply CHANGES.current:12]: Also "convertion" (should be "conversion").

And the convention is to "tag" the entry with the language it applies to like so:

```
2023-12-01: saiarcot895
            [Python] #2413 Prevent potential multi-threading crash; gracefully exit running
            daemon threads on main thread exit.
```

File: Examples/test-suite/overload_simple.i [hunk 6]
```diff
@@ -166,10 +166,21 @@ char *fid(double, double) {
...
 %inline %{
-unsigned long long ull() { return 0ULL; }
-unsigned long long ull(unsigned long long ull) { return ull; }
-long long ll() { return 0LL; }
-long long ll(long long ull) { return ull; }
+#include <errno.h>
+#include <math.h>
+
+int sizeof_long() { return sizeof(long); }
+double getnextafter(double from, double to) { double ret = nextafter(from, to); errno = 0; return ret; }
```
Comment (ojwb) [inline Examples/test-suite/overload_simple.i:173]: `nextafter()` is C99 so this testcase may fail if using a C90 compiler which is still the baseline we aim to support.

I don't think we have a "C99-only testcase" mechanism, but `nextafter()` is also C++11 and we do have a "C++11-only testcase" mechanism so this could be made a C++11 testcase (rename to add a `cpp11_` prefix and move it to `CPP11_TEST_CASES`).  Or just drop this helper and skip the parts of the `_runme.py` that requires it if using Python < 3.9.

Comment (degasus) [reply Examples/test-suite/overload_simple.i]: I've dropped this C helpers and added an optional fallback to numpy instead. If both is missing, the tests will be skipped. I hope the coverage is still good enough.
Our behavior changed with Python >= 3.10, so having at least 3.9 should still test all required cases.

File: CHANGES.current (hunk 5 above)
Comment (degasus) [reply CHANGES.current]: Done.
Nitpicking: Which convention should I follow for the indentation? 12 spaces or one tab + 4 spaces?

And as usual, sorry for my bad English ;-)

File: Lib/python/pyprimtypes.swg (hunk 2 above)
Comment (degasus) [reply Lib/python/pyprimtypes.swg:108]: I sadly still have no better option than hardcoding this constants in the generated C code. And yes, this is hardcoded for IEEE floats only. Do you want me to add some macros to check if the predefined macros in float.h matches our expections and #error else?

However I skipped your suggestion for generating `llmaxd`. This is within the test, and for the cast_mode feature, the test is currently more a documentation of the behavior than checking for the designed behavior. I fear that importing the size of floats from C makes this just more unreadable.

**Comment by degasus** [comment]
> I've added it to the 4.2 milestone for consideration (but note that's not a guarantee it'll get merged for 4.2.0).

Thanks a lot. Obviously no guarantee for anything, but this makes it less likely to just be forgotten :-)

**Comment by ojwb** [comment]
> Note: The failing test also fails on master.

It's marked as "can fail" (it's due to a bug in upstream ruby which has been fixed and we're just waiting for that fix to percolate through to the ruby 3.3 development builds we're using in CI).

It shows as green in https://github.com/swig/swig/actions but github is stupid and still treats that as a failing test in PRs.

File: CHANGES.current (hunk 5 above)
Comment (ojwb) [reply CHANGES.current]: SWIG coding style expects a tab to move to the next multiple of 8 column so they're equivalent.

File: Examples/test-suite/python/python_overload_simple_cast_runme.py (hunk 1 above)
Comment (wsfulton) [reply Examples/test-suite/python/python_overload_simple_cast_runme.py:168]: Agreed, `9223372036854775807 == 0x7fffffffffffffff`

File: Examples/test-suite/overload_simple.i (hunk 3 above)
Comment (wsfulton) [reply Examples/test-suite/overload_simple.i:174]: Hard to tell from the diffs displayed here, but `ull()` is now `as_ull()` and `ull()` is now `as_ll()`.

File: Lib/python/pyprimtypes.swg (hunk 2 above)
Comment (wsfulton) [reply Lib/python/pyprimtypes.swg:108]: Agreed that hard coding is not ideal, but given it is not easily calculated portably, I've put in a comment on how to calculate it as the next best thing. See eac27f7d3ecfe0c120f202eb1a1da081060884f6.

File: Examples/test-suite/python/python_overload_simple_cast_runme.py [hunk 7]
```diff
@@ -164,34 +164,129 @@ def __float__(self):
...
+lmaxd = float(2**31 - 1)
+lmind = float(-2**31)
+lmaxd_v = lmaxd # expected value after the cast
+lmind_v = lmind
+if hexversion < 0x30a0000:
+    # PyLong_AsLong(float) truncated the input before 3.10
+    lmaxd = nextafter(float(2**31), 0.0)
+    lmind = nextafter(float(-2**31 - 1), 0.0)
+    lmaxd_v = float(2**31 - 1)
+    lmind_v = float(-2**31)
```
Comment (wsfulton) [inline Examples/test-suite/python/python_overload_simple_cast_runme.py:235]: I ran the tests on 32-bit Python < 3.10 to see that this change is required, but I didn't figure out why. Do you have a link to the exact release note that fixes this truncation problem?

**Comment by wsfulton** [comment]
Thanks for this pull request @degasus . It took a while digesting and checking it, it must have taken you even longer to figure out the problem and fix. There's just one comment on the 32-bit Python if you can take a look.

File: Examples/test-suite/python/python_overload_simple_cast_runme.py [hunk 8]
```diff
@@ -164,34 +164,129 @@ def __float__(self):
...
+exceptTypeError(as_l, lmax + 1, "as_l(lmax + 1)")
+exceptTypeError(as_l, nextafter(lmind, -inf), "as_l(lmind - LSB)")
+exceptTypeError(as_l, nextafter(lmaxd, inf), "as_l(lmaxd + LSB)")
+
 # unsigned long long
-ullmax = 9223372036854775807  # 0xffffffffffffffff
-ullmaxd = 9007199254740992.0
+ullmax = 2**64 - 1
 ullmin = 0
+ullmaxd = float(2**53) # 64 bit double significand
```
Comment (wsfulton) [inline Examples/test-suite/python/python_overload_simple_cast_runme.py:257]: Also, why test 2**53 for `unsigned long long`, this surely should be `nextafter(float(2**64), 0.0)`, that is, the same as 64-bit `unsigned long` as `sizeof(unsigned long) == sizeof(unsigned long long)` on 64-bit platforms.

File: Examples/test-suite/python/python_overload_simple_cast_runme.py (hunk 7 above)
Comment (degasus) [reply Examples/test-suite/python/python_overload_simple_cast_runme.py:235]: https://docs.python.org/3/c-api/long.html#c.PyLong_AsLong
Before 3.10, `__int__` was used and it truncated any float. Since 3.10, it only tries to use `__index__`, which fails for float. I think we might want to call it a bug to use `PyLong_AsLong` for a float, but this defines the current behavior: https://github.com/swig/swig/blob/master/Lib/python/pyprimtypes.swg#L97

File: Examples/test-suite/python/python_overload_simple_cast_runme.py (hunk 8 above)
Comment (degasus) [reply Examples/test-suite/python/python_overload_simple_cast_runme.py:257]: I think this behavior is implemented on purpose [here](https://github.com/swig/swig/blob/master/Lib/python/pyprimtypes.swg#L291) or [here](https://github.com/swig/swig/blob/master/Lib/perl5/perlprimtypes.swg#L232). By the way, this reminds me to double check the behavior on perl if it has the same bug as fixed here...

cast_mode (sometimes) check for how many LSBs the number is off, and it (sometimes) check for 53b precision. In my opinion, there are reasons to do so and reasons not to do so. I just dislike the "sometimes" based on plattforms. If you have an opinion on which behavior you like more, I will be very happily rewrite #2097 to actually get the same behavior on all plattforms. But I think this should not be considered for the current release.

**Comment by degasus** [comment]
> Thanks for this pull request @degasus . It took a while digesting and checking it, it must have taken you even longer to figure out the problem and fix. There's just one comment on the 32-bit Python if you can take a look.

Thanks a lot for the review, and marry x-mas. What a honor to get your attention at this time!

Figuring out the problem was actually easy, clang provided a warning that the constant LONG_MAX casted to double does not exist. Figuring out the behavior on all plattforms however took a while ;-)

File: Examples/test-suite/python/python_overload_simple_cast_runme.py (hunk 8 above)
Comment (degasus) [reply Examples/test-suite/python/python_overload_simple_cast_runme.py:257]: ..... Sorry for missing to check perl ....  https://github.com/swig/swig/blob/master/Lib/perl5/perlprimtypes.swg#L90 and https://github.com/swig/swig/blob/master/Lib/perl5/perlprimtypes.swg#L157

Comment (wsfulton) [reply Examples/test-suite/python/python_overload_simple_cast_runme.py:257]: > rewrite https://github.com/swig/swig/pull/2097 to actually get the same behavior on all plattforms.

Yes, so let's make consistent and resolve in the rewrite of #2097 and agreed it is not going to make it into swig-4.2.0 due out in a few days.
